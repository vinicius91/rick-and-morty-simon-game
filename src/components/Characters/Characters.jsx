// @flow

import React, { Component } from 'react';
import { Wrapper } from './Characters.styles';
import Character from './Character/Character';
import type { Characters } from '../../types/character';

type ClickHandlerFunction = (id: number) => void;

type Color = {
  base: string,
  dark: string
};

type Colors = Array<Color>;

type CharactersProps = {
  characters: Characters,
  clickCharacter: ClickHandlerFunction
};

const simonColors: Colors = [
  {
    base: '#5fd382',
    dark: '#4ca868'
  },
  {
    base: '#dc5245',
    dark: '#b04137'
  },
  {
    base: '#e6e564',
    dark: '#cfce5a'
  },
  {
    base: '#038fe4',
    dark: '#0272b6'
  }
];

const renderCharacters = (
  characters: Characters,
  colors: Colors,
  handler: ClickHandlerFunction
) => {
  if (
    characters &&
    simonColors &&
    characters.length > 0 &&
    simonColors.length > 0
  ) {
    return characters.map((character, i) => (
      <Character
        imageUrl={character.imageUrl}
        id={character.id}
        key={character.id}
        color={simonColors[i].base}
        colorDark={simonColors[i].dark}
        active={false}
        characterClickHandler={handler}
      />
    ));
  }

  return '';
};

class CharactersComponent extends Component {
  componentDidMount() {
    this.props.fetchCharacters();
  }

  render() {
    const { characters, clickCharacter } = this.props;
    return (
      <Wrapper>
        {renderCharacters(characters, simonColors, clickCharacter)}
      </Wrapper>
    );
  }
}

export default CharactersComponent;
