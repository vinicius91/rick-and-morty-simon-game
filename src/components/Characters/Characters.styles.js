import styled from 'styled-components';

export const Wrapper = styled.div`
  width: 100%;
  flex-grow: 1;
  display: flex;
  flex-flow: row wrap;
  justify-content: space-around;
`;
