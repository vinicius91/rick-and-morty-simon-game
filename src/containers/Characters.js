import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { clickCharacter } from '../store/actions/game';
import { fetchCharacters } from '../store/actions/character';
import CharactersComponent from '../components/Characters/Characters';

const mapStateToProps = state => {
  console.log(state);

  return state.characters;
};

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      clickCharacter,
      fetchCharacters
    },
    dispatch
  );

const CharactersContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(CharactersComponent);

export default CharactersContainer;
