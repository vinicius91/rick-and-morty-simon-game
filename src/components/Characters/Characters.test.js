import React from 'react';
import renderer from 'react-test-renderer';
import Characters from './Characters';

describe('Characters Component', () => {
  const charactersArray = [
    {
      id: 1,
      imageUrl: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg'
    },
    {
      id: 2,
      imageUrl: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg'
    },
    {
      id: 3,
      imageUrl: 'https://rickandmortyapi.com/api/character/avatar/3.jpeg'
    },
    {
      id: 4,
      imageUrl: 'https://rickandmortyapi.com/api/character/avatar/4.jpeg'
    }
  ];

  const fixture = {
    characters: charactersArray,
    loading: false,
    isUserStep: false,
    activeCharacter: 0,
    clickCharacter: () => {},
    fetchCharacters: () => {}
  };

  it('Should render with 4 characters', () => {
    const component = renderer.create(<Characters {...fixture} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render with loading', () => {
    const loadingFixture = { ...fixture, loading: true };
    const component = renderer.create(<Characters {...loadingFixture} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render without characters', () => {
    const loadingFixture = { ...fixture, characters: [] };
    const component = renderer.create(<Characters {...loadingFixture} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
