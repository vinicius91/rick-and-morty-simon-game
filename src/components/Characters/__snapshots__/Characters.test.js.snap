// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Characters Component Should render with 4 characters 1`] = `
<div
  className="Charactersstyles__Wrapper-sc-1z0bbli-0 kdlhnX"
>
  <div
    className="Characterstyles__RegularCharacter-sc-77sfuh-0 jKryPT"
    color="#5fd382"
    onClick={[Function]}
  >
    <audio
      src="https://simongame.s3.amazonaws.com/simonSound1.mp3"
    />
  </div>
  <div
    className="Characterstyles__RegularCharacter-sc-77sfuh-0 iwLkVP"
    color="#dc5245"
    onClick={[Function]}
  >
    <audio
      src="https://simongame.s3.amazonaws.com/simonSound2.mp3"
    />
  </div>
  <div
    className="Characterstyles__RegularCharacter-sc-77sfuh-0 iqtftX"
    color="#e6e564"
    onClick={[Function]}
  >
    <audio
      src="https://simongame.s3.amazonaws.com/simonSound3.mp3"
    />
  </div>
  <div
    className="Characterstyles__RegularCharacter-sc-77sfuh-0 dmXYji"
    color="#038fe4"
    onClick={[Function]}
  >
    <audio
      src="https://simongame.s3.amazonaws.com/simonSound4.mp3"
    />
  </div>
</div>
`;

exports[`Characters Component Should render with loading 1`] = `
<div
  className="Charactersstyles__Wrapper-sc-1z0bbli-0 kdlhnX"
>
  <div
    className="Loadingstyles__Wrapper-sc-1sj5pia-0 doNlxm"
  >
    <div
      className="Loadingstyles__Portal-sc-1sj5pia-1 gPsram"
    >
      <div
        className="Loadingstyles__Swish-sc-1sj5pia-10 fnDiug"
      >
        <span
          className="Loadingstyles__SwishSpan-sc-1sj5pia-11 Loadingstyles__SwishSpanOne-sc-1sj5pia-12 isPzsW"
        />
        <span
          className="Loadingstyles__SwishSpan-sc-1sj5pia-11 Loadingstyles__SwishSpanTwo-sc-1sj5pia-13 kykqkM"
        />
        <span
          className="Loadingstyles__SwishSpan-sc-1sj5pia-11 Loadingstyles__SwishSpanThree-sc-1sj5pia-14 kBSnuF"
        />
        <span
          className="Loadingstyles__SwishSpan-sc-1sj5pia-11 Loadingstyles__SwishSpanFour-sc-1sj5pia-15 dMZCuu"
        />
      </div>
      <div
        className="Loadingstyles__Swirl-sc-1sj5pia-2 jkzykY"
      >
        <span
          className="Loadingstyles__SwirlSpan-sc-1sj5pia-6 Loadingstyles__SwirlSpanOne-sc-1sj5pia-7 iBNFKm"
        />
        <span
          className="Loadingstyles__SwirlSpan-sc-1sj5pia-6 Loadingstyles__SwirlSpanTwo-sc-1sj5pia-8 hSHYsR"
        />
        <span
          className="Loadingstyles__SwirlSpan-sc-1sj5pia-6 Loadingstyles__SwirlSpanThree-sc-1sj5pia-9 ligCez"
        />
        <div
          className="Loadingstyles__Swirl-sc-1sj5pia-2 Loadingstyles__SwirlDouble-sc-1sj5pia-3 dqKKzo"
        >
          <span
            className="Loadingstyles__SwirlSpan-sc-1sj5pia-6 Loadingstyles__SwirlSpanOne-sc-1sj5pia-7 iBNFKm"
          />
          <div
            className="Loadingstyles__Swirl-sc-1sj5pia-2 Loadingstyles__SwirlTriple-sc-1sj5pia-4 kPttOI"
          >
            <span
              className="Loadingstyles__SwirlSpan-sc-1sj5pia-6 Loadingstyles__SwirlSpanOne-sc-1sj5pia-7 iBNFKm"
            />
            <div
              className="Loadingstyles__Swirl-sc-1sj5pia-2 Loadingstyles__SwirlTriple-sc-1sj5pia-4 kPttOI"
            >
              <span
                className="Loadingstyles__SwirlSpan-sc-1sj5pia-6 Loadingstyles__SwirlSpanOne-sc-1sj5pia-7 iBNFKm"
              />
              <div
                className="Loadingstyles__Swirl-sc-1sj5pia-2 Loadingstyles__SwirlTriple-sc-1sj5pia-4 kPttOI"
              >
                <span
                  className="Loadingstyles__SwirlSpan-sc-1sj5pia-6 Loadingstyles__SwirlSpanOne-sc-1sj5pia-7 iBNFKm"
                />
                <div
                  className="Loadingstyles__Swirl-sc-1sj5pia-2 Loadingstyles__SwirlSpecial-sc-1sj5pia-5 ifLXZo"
                >
                  <span
                    className="Loadingstyles__SwirlSpan-sc-1sj5pia-6 Loadingstyles__SwirlSpanOne-sc-1sj5pia-7 iBNFKm"
                  />
                  <div
                    className="Loadingstyles__Swirl-sc-1sj5pia-2 Loadingstyles__SwirlTriple-sc-1sj5pia-4 kPttOI"
                  >
                    <span
                      className="Loadingstyles__SwirlSpan-sc-1sj5pia-6 Loadingstyles__SwirlSpanOne-sc-1sj5pia-7 iBNFKm"
                    />
                    <div
                      className="Loadingstyles__Swirl-sc-1sj5pia-2 Loadingstyles__SwirlSpecial-sc-1sj5pia-5 ifLXZo"
                    >
                      <span
                        className="Loadingstyles__SwirlSpan-sc-1sj5pia-6 Loadingstyles__SwirlSpanOne-sc-1sj5pia-7 iBNFKm"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Characters Component Should render without characters 1`] = `
<div
  className="Charactersstyles__Wrapper-sc-1z0bbli-0 kdlhnX"
>
  
</div>
`;
